crates/js-sys/src/lib.rs

```diff
@@ -1174,6 +1174,20 @@ extern {
     pub fn next(this: &Iterator) -> Result<IteratorNext, JsValue>;
 }
 
+impl std::iter::Iterator for Iterator {
+    type Item = JsValue;
+
+    fn next(&mut self) -> Option<JsValue> {
+        let val = Iterator::next(&*self)
+            .expect("JS object didn't obey iterator protocol");
```

Comment:
Iterables (such as generators) can totally throw, and it isn't really violating the protocol :-/

![generator-throws](https://user-images.githubusercontent.com/74571/45232423-333f1700-b284-11e8-9c1c-301b727f2612.png)

I think we should have `Iterator<Item = Result<JsValue, JsValue>>` and then fuse the iterator such that calling next after the first `Err` always returns `None` and doesn't call back into JS again.

Reply:
Interesting! 

I don't think we could auto-fuse though unless we start setting a JS property on the iterator (which seems icky) or have a wrapper in Rust. I'd be ok returning a `Result<JsValue, JsValue>` but that would make fusing difficult for sure

Reply:
I think the fusing is important for combinators and `.collect()` to work correctly.

Perhaps we should implement `IntoIterator` for `&'a js_sys::Iterator` witha  custom type that has a `bool` for fusing?